code/dictation.py

```diff
@@ -143,41 +171,98 @@ def __init__(self):
         self.reset()
 
     def reset(self):
+        self.reset_except_forced()
+        self.force_no_space = False
+        self.force_cap = False
+        self.force_no_cap = False
+
+    def reset_except_forced(self):
         self.before = ""
-        self.state = "sentence start"
+        self.state = None
 
     def update_context(self, before):
         if before is None: return
-        self.reset()
+        self.reset_except_forced()
         self.pass_through(before)
 
     def pass_through(self, text):
         _, self.state = auto_capitalize(text, self.state)
         self.before = text or self.before
 
-    def format(self, text):
-        if needs_space_between(self.before, text):
+    def format(self, text, auto_cap=True):
+        if not self.force_no_space and needs_space_between(self.before, text):
             text = " " + text
-        text, self.state = auto_capitalize(text, self.state)
+        self.force_no_space = False
+        if auto_cap:
+            text, self.state = auto_capitalize(text, self.state)
+        if self.force_cap:
+            text = format_first_letter(text, lambda s: s.capitalize())
+            self.force_cap = False
+        if self.force_no_cap:
+            text = format_first_letter(text, lambda s: s.lower())
+            self.force_no_cap = False
```

Comment:
Originally I did in fact reuse self.state. The problem is it's nice to override default capitalization (e.g. make proper names lowercase) when desired, so I needed the ability to _force_, which is more than self.state offered. In terms of separation of concerns, I think it's cleaner to have separate state instance variables instead of trying to package everything in one. I agree there is some duplication between format_first_letter and auto_capitalize, though, although I'm not sure trying to do all that in one function is going to simplify things.

Reply:
Yeah, I agree.